### PR TITLE
Introduce Try Clear Registration; Delay activation

### LINKED ISSUE #916: More clarity around waiting for an active worker to finish
I'm working on a Chrome [bug](https://bugs.chromium.org/p/chromium/issues/detail?id=616331) where we're immediately terminating the incumbent worker during activation instead of waiting for it to finish its "in-progress requests" as the spec says. Since update occurs soon after page load, with skipWaiting() we run into this issue in the wild.

I'm wondering though, what exactly is a "request". Is this referring to fetch event only or is it also intended to mean functional events?

Elsewhere the spec also mentions "in-flight requests" which are "triggered by" a worker, which seems a different thing that "in-progress requests" being handled by a worker. 

The spec also doesn't really mention what to do with new incoming requests: do they go to the incumbent worker or the soon-to-be activating worker?

My current plan is to:
- Wait for all events already dispatched to the incumbent worker to finish (subject to our usual timeout)
- All new events are put on hold until the soon-to-be activating worker is activated

How does Firefox deal with this issue?

### PR DESCRIPTION
**Pull request by jungkees**
This is a follow-up change, to [1], which addresses the issues posed in
[2] and the comments afterwards.

This change makes the callers to the Clear Registration algorithm to
check if no client is using the registration and the registration's
associated service workers have no pending extended events before
invoking the algorithm. Unregister, Handle Service Worker Client Unload,
and the microtasks queued by the extension promises' settlement handlers
invoke Try Clear Registration.

Other error handling steps on registrations that have no associated
worker and Handle User Agent Shutdown invoke Clear Registration directly
as they do.

[1] https://github.com/w3c/ServiceWorker/commit/f1a1120ff07120a2e8f311fb61f429c9a74db1ef
[2] https://github.com/w3c/ServiceWorker/issues/916#issuecomment-237272060

Summary of the issues: https://github.com/w3c/ServiceWorker/issues/916#issuecomment-281274267.

Fixes #916.

### CODE REVIEW

**Comment by jungkees** [comment]
@jakearchibald, @mkruisselbrink, This PR is to address https://github.com/w3c/ServiceWorker/issues/916#issuecomment-281274267. PTAL.

/cc @wanderview @mattto @annevk @slightlyoff

**Comment by jungkees** [comment]
I left Handle User Agent Shutdown unchanged as I think it shouldn't miss the chance to update the registrations (clear and activate operations).

File: docs/index.bs [hunk 1]
```diff
@@ -1323,7 +1323,9 @@ spec: webappsec-referrer-policy; urlPrefix: https://w3c.github.io/webappsec-refe
 
         1. Upon [=upon fulfillment|fulfillment=] or [=upon rejection|rejection=] of |f|, [=queue a microtask=] to run these substeps:
             1. Decrement the [=ExtendableEvent/pending promises count=] by one.
-            1. Invoke [=Try Activate=] with the [=context object=]'s [=relevant global object=]'s associated [=ServiceWorkerGlobalScope/service worker=]'s [=containing service worker registration=].
+            1. Let |registration| be the [=context object=]'s [=relevant global object=]'s associated [=ServiceWorkerGlobalScope/service worker=]'s [=containing service worker registration=].
+            1. If |registration|'s [=uninstalling flag=] is set, invoke [=Try Clear Registration=] with |registration|.
```
Comment (jakearchibald) [inline docs/index.bs]: Should this check to see if the registration is not null before checking the flag?

Comment (jungkees) [inline docs/index.bs]: I think the event has been alive up to this point so no Try Clear Registration should have cleared the registration yet. So, it seems we can assert that the registration must not be null here.

**Comment by jungkees** [comment]
@jakearchibald, thanks for the review!

**Pull request merged**